WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,350 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Client.CredentialDependencies
+{
+	public class DependencyGraph
+	{
+		public const int K = ProtocolConstants.CredentialNumber;
+
+		private DependencyGraph(IEnumerable<IEnumerable<long>> initialValues)
+		{
+			var initialValuesImmutable = initialValues.Select(x => x.ToImmutableArray()).ToImmutableArray();
+			if (initialValuesImmutable.Any(x => x.Length != (int)CredentialType.NumTypes))
+			{
+				throw new ArgumentException($"Number of credential values must be {CredentialType.NumTypes}");
+			}
+			if (initialValuesImmutable.Any(x => x.Where(y => y != 0).Select(y => y.CompareTo(0)).Distinct().Count() != 1))
+			{
+				throw new ArgumentException($"Credential values of a node must all have the same sign.");
+			}
+
+			for (CredentialType i = 0; i < CredentialType.NumTypes; i++)
+			{
+				if (initialValuesImmutable.Sum(x => x[(int)i]) < 0)
+				{
+					throw new ArgumentException("Overall balance must not be negative");
+				}
+			}
+
+			// per node entries created in AddNode, querying nodes not in the
+			// graph should result in key errors.
+
+			Successors = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, HashSet<CredentialDependency>>()).ToImmutableArray();
+			Predecessors = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, HashSet<CredentialDependency>>()).ToImmutableArray();
+			EdgeBalances = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, long>()).ToImmutableArray();
+
+			Vertices = ImmutableList<RequestNode>.Empty;
+
+			foreach (var node in initialValuesImmutable.Select((values, i) => new RequestNode(i, values)))
+			{
+				// enforce at least one value != 0? all values? it doesn't actually matter for the algorithm
+				AddNode(node);
+			}
+		}
+
+		public ImmutableList<RequestNode> Vertices { get; private set; }
+
+		// Internal properties used to keep track of effective values and edges
+		private ImmutableArray<Dictionary<int, long>> EdgeBalances { get; }
+		private ImmutableArray<Dictionary<int, HashSet<CredentialDependency>>> Successors { get; }
+		private ImmutableArray<Dictionary<int, HashSet<CredentialDependency>>> Predecessors { get; }
+
+		// TODO doc comment
+		// Public API: construct a graph from amounts, and resolve the
+		// credential dependencies. Should only produce valid graphs.
+		// IDs are positive ints assigned in the order of the enumerable, but
+		// Vertices will contain more elements if there are reissuance nodes.
+		public static DependencyGraph ResolveCredentialDependencies(IEnumerable<IEnumerable<long>> amounts)
```

Comment:
Why do I have a sublist? I thought we are just giving this the desired amounts, that I want to have 0.5 1 and 2 at the end.

Reply:
Maybe more comment or renaming amounts can help.

Reply:
:+1: on renaming, there's one value for each credential type, so this should probably be called `values` and i'll make it into a proper doc comment